aten/src/ATen/core/dynamic_type.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <memory>
+
+#include <ATen/core/class_type.h>
+#include <ATen/core/jit_type_base.h>
+#include <c10/util/Optional.h>
+
+using DynamicTypeBits = std::uint32_t;
+#define DYNAMIC_TYPE_BIT(x) (1u << x)
+
+constexpr DynamicTypeBits kDynamicCovariantTypeBit = DYNAMIC_TYPE_BIT(31);
+constexpr DynamicTypeBits kDynamicAnyTypeBit = DYNAMIC_TYPE_BIT(30);
+constexpr DynamicTypeBits kDynamicSingletonTypeBit = DYNAMIC_TYPE_BIT(29);
+
+constexpr DynamicTypeBits kDynamicNoneTypeBit = DYNAMIC_TYPE_BIT(1);
+constexpr DynamicTypeBits kDynamicIntTypeBit = DYNAMIC_TYPE_BIT(3);
+constexpr DynamicTypeBits kDynamicFloatTypeBit = DYNAMIC_TYPE_BIT(4);
+constexpr DynamicTypeBits kDynamicComplexTypeBit = DYNAMIC_TYPE_BIT(5);
+constexpr DynamicTypeBits kDynamicListTypeBit = DYNAMIC_TYPE_BIT(7);
+constexpr DynamicTypeBits kDynamicTupleTypeBit = DYNAMIC_TYPE_BIT(8);
+
+#define FORALL_DYNAMIC_TYPES(_)                                              \
+  _(Tensor, DYNAMIC_TYPE_BIT(0))                                             \
+  _(None, kDynamicNoneTypeBit)                                               \
+  _(Bool, DYNAMIC_TYPE_BIT(2))                                               \
+  _(Int, kDynamicIntTypeBit)                                                 \
+  _(Float, kDynamicFloatTypeBit)                                             \
+  _(Complex, kDynamicComplexTypeBit)                                         \
+  _(Number,                                                                  \
+    (kDynamicIntTypeBit | kDynamicFloatTypeBit | kDynamicComplexTypeBit))    \
+  _(String, DYNAMIC_TYPE_BIT(6))                                             \
+  _(List, kDynamicListTypeBit)                                               \
+  _(Tuple, (kDynamicTupleTypeBit | kDynamicCovariantTypeBit))                \
+  _(Dict, DYNAMIC_TYPE_BIT(9))                                               \
+  _(Class, DYNAMIC_TYPE_BIT(10))                                             \
+  _(Optional,                                                                \
```

Comment:
How do we handle `Union` in the future if it is needed on mobile? say in an `isinstance` check? Should we canonicalize `Optional` to `Union` in that case to simplify subtype check?

Reply:
Union type is not needed at **runtime**, because it only makes sense to use it in type annotations, right now both schema and `isinstance` won't support UnionType. This is one example something is useful in compilation but not in runtime.